runtime/src/main/java/org/corfudb/runtime/collections/StreamManager.java

```diff
@@ -0,0 +1,261 @@
+            throw new StreamSubscriptionException(
+                    "StreamManager:subscriber already registered "
+                            + streamListener + ". Maybe call unregister first?");
+        }
+        log.info("StreamManager::subscribe {}, startAddress {}, namespace {}, tables {}",
+                streamListener, startAddress, namespace, tablesOfInterest.toString());
+        SubscriberTask task = new SubscriberTask(this,
+                streamListener, namespace, tablesOfInterest, startAddress);
+        final ScheduledFuture<?> scheduledFuture = scheduledExecutorService.scheduleWithFixedDelay(
+                task, 0, DEFAULT_SLEEP_MILLIS, TimeUnit.MILLISECONDS);
+        subscriptions.put(streamListener, new StreamSubscriber(task, scheduledFuture));
+    }
+
+    /**
+     * Unsubscribe a prior subscription.
+     *
+     * @param streamListener Client listener.
+     */
+    public void unsubscribe(@Nonnull StreamListener streamListener) {
+        StreamSubscriber streamSubscriber = unsubscribeInternal(streamListener);
+        if (streamSubscriber != null) {
+            streamSubscriber.getScheduledFuture().cancel(false);
+        }
+    }
+
+    /**
+     * Internal method that can be safely invoked from the same thread executing the task.
+     * @param streamListener client's listener object.
+     * @return - returns the stream subscriber context for lifecycle management.
+     */
+    private synchronized StreamSubscriber unsubscribeInternal(@Nonnull StreamListener streamListener) {
+        StreamSubscriber subscriber = subscriptions.remove(streamListener);
+        if (subscriber == null) {
+            log.warn("StreamManager::unsubscribe has no context for {}", streamListener.toString());
+            return null;
+        }
+        // Since we are not force interrupting the thread, set a flag for it to shutdown
+        subscriber.getSubscriberTask().isShutdown.set(true);
+        log.info("Unsubscribed StreamListener {}", streamListener.toString());
+        runtime.getStreamsView().removeStream(subscriber.getSubscriberTask().txnStream);
+        return subscriber;
+    }
+
+    class SubscriberTask implements Runnable {
+        /**
+         * The callback context
+         */
+        private final StreamListener listener;
+
+        /**
+         * Parent class to manage errors.
+         */
+        @Getter
+        private final StreamManager streamManager;
+
+        /**
+         * Namespace of interest.
+         */
+        @Getter
+        private final String namespace;
+
+        @Setter
+        private final AtomicBoolean isShutdown = new AtomicBoolean(false);
+
+        /**
+         * Tables of interest.
+         *
+         * Map of (Stream Id - TableSchema)
+         */
+        @Getter
+        private final Map<UUID,
+                TableSchema<? extends Message, ? extends Message, ? extends Message>> tablesOfInterest;
+
+        public AtomicBoolean shutdown;
```

Comment:
There is already a `isShutdown` variable on line 168 right?

Reply:
fixed. thanks